src/main/scala/strawman/collection/Set.scala

```diff
@@ -42,9 +58,13 @@ trait SetLike[A, +C[X] <: Set[X]]
 trait SetMonoTransforms[A, +Repr]
   extends IterableMonoTransforms[A, Repr] {
 
-  def & (that: Set[A]): Repr
+  def & (that: Set[A]): Repr = this.filter(that)
+
+}
+
+trait SetPolyTransforms[A, +C[X]] extends IterablePolyTransforms[A, C] {
 
-  def ++ (that: Set[A]): Repr
+  def ++ (that: Set[A]): C[A]
```

Comment:
Actually, I think this is not possible to implement this `++` signature in a sound way: see my attempt here: https://github.com/scala/collection-strawman/compare/hashet-benchmarks...unsound-set-concat

Let me explain a bit better the tradeoff. The `++` method inherited from `Iterable` has the following signature:

~~~ scala
def ++ [B >: A](that: IterableOnce[B]): Set[B]
~~~

While the `Set`-specific version has the following signature:

~~~ scala
def ++ (that: Set[A]): Set[A]
~~~

The former forces us to rebuild a whole `Set[B]` and to add to it all the elements of `this` and all the elements of `that`. The latter allows us to reuse `this` as it is and just add to it all the elements of `that`. This optimization can not be achieved with the former type signature because we would have to upcast `this` into a `Set[B]`, which is unsound because of the way `contains` is defined (`A` is in contravariant position). By the way, this makes me think that we will have the same issue with `contains`, if we define it in `Iterable`.

I would be tempted to remove these operations from `Iterable` but that would make `Iterable` less useful… Alternatively, we could live with overloads:

![image](https://cloud.githubusercontent.com/assets/332812/24500211/c922806a-1544-11e7-8b5e-6f654b13f7de.png)